File: components/select-combobox/flutter/ciata_select.dart
import 'package:flutter/material.dart';

class CiataSelectOption<T> {
  CiataSelectOption({
    required this.value,
    required String label,
    this.enabled = true,
  }) : label = label.trim() {
    if (this.label.isEmpty) {
      throw ArgumentError.value(label, 'label', 'não pode ser vazio');
    }
  }

  final T value;
  final String label;
  final bool enabled;
}

/// Implementação experimental do CMP-0005 Select para Flutter.
class CiataSelect<T> extends StatelessWidget {
  CiataSelect({
    super.key,
    required String label,
    required this.options,
    required this.value,
    required this.onChanged,
    this.requiredField = false,
    this.enabled = true,
    String? helpText,
    String? errorText,
  })  : label = label.trim(),
        helpText = helpText?.trim(),
        errorText = errorText?.trim() {
    if (this.label.isEmpty) {
      throw ArgumentError.value(label, 'label', 'não pode ser vazio');
    }
    if (options.isEmpty) {
      throw ArgumentError('options não pode ser vazio');
    }
    final values = options.map((option) => option.value).toList(growable: false);
    if (values.toSet().length != values.length) {
      throw ArgumentError('Valores das opções devem ser únicos');
    }
    if (value != null && !values.contains(value)) {
      throw ArgumentError.value(value, 'value', 'não pertence às opções do Select');
    }
  }

  final String label;
  final List<CiataSelectOption<T>> options;
  final T? value;
  final ValueChanged<T?>? onChanged;
  final bool requiredField;
  final bool enabled;
  final String? helpText;
  final String? errorText;

  @override
  Widget build(BuildContext context) {
    final visibleLabel = requiredField ? '$label (obrigatório)' : label;

    return DropdownButtonFormField<T>(
      value: value,
      onChanged: enabled ? onChanged : null,
      decoration: InputDecoration(
        labelText: visibleLabel,
        helperText: errorText?.isNotEmpty == true ? null : helpText,
        errorText: errorText?.isNotEmpty == true ? errorText : null,
      ),
      items: options
          .map(
            (option) => DropdownMenuItem<T>(
              value: option.value,
              enabled: enabled && option.enabled,
              child: Text(option.label),
            ),
          )
          .toList(),
    );
  }
}

File: components/select-combobox/android/CiataSelect.kt
package org.ciata.ds.select

import androidx.compose.foundation.layout.Column
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics

data class CiataSelectOption(
    val value: String,
    val label: String,
    val enabled: Boolean = true,
)

/** Implementação experimental do CMP-0005 Select para Jetpack Compose. */
@Composable
fun CiataSelect(
    label: String,
    options: List<CiataSelectOption>,
    selectedValue: String?,
    onSelected: (String) -> Unit,
    modifier: Modifier = Modifier,
    required: Boolean = false,
    enabled: Boolean = true,
    helpText: String? = null,
    errorText: String? = null,
) {
    val normalizedLabel = label.trim()
    val normalizedOptions = options.map { option -> option.copy(label = option.label.trim()) }
    val normalizedHelpText = helpText?.trim()
    val normalizedErrorText = errorText?.trim()

    require(normalizedLabel.isNotEmpty()) { "label não pode ser vazio." }
    require(normalizedOptions.isNotEmpty()) { "options não pode ser vazio." }
    require(normalizedOptions.all { it.label.isNotEmpty() }) { "Rótulos das opções não podem ser vazios." }
    require(normalizedOptions.map { it.value }.distinct().size == normalizedOptions.size) { "Valores devem ser únicos." }
    require(selectedValue == null || normalizedOptions.any { it.value == selectedValue }) {
        "selectedValue não pertence às opções do Select."
    }

    var expanded by remember { mutableStateOf(false) }
    val selectedLabel = normalizedOptions.firstOrNull { it.value == selectedValue }?.label.orEmpty()
    val semanticModifier = if (!normalizedErrorText.isNullOrEmpty()) {
        modifier.semantics { error(normalizedErrorText) }
    } else modifier

    Column(modifier = semanticModifier) {
        ExposedDropdownMenuBox(
            expanded = expanded,
            onExpandedChange = { if (enabled) expanded = !expanded },
        ) {
            TextField(
                value = selectedLabel,
                onValueChange = {},
                readOnly = true,
                enabled = enabled,
                label = { Text(if (required) "$normalizedLabel (obrigatório)" else normalizedLabel) },
                trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded) },
                modifier = Modifier.menuAnchor(),
                isError = !normalizedErrorText.isNullOrEmpty(),
            )

            ExposedDropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false },
            ) {
                normalizedOptions.forEach { option ->
                    DropdownMenuItem(
                        text = { Text(option.label) },
                        enabled = enabled && option.enabled,
                        onClick = {
                            onSelected(option.value)
                            expanded = false
                        },
                    )
                }
            }
        }

        if (!normalizedHelpText.isNullOrEmpty()) Text(normalizedHelpText)
        if (!normalizedErrorText.isNullOrEmpty()) Text(normalizedErrorText)
    }
}
